File: hdf-converters/shared/go/anchor_test.go
package shared

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestCountXMLElements(t *testing.T) {
	xml := []byte(`<Benchmark><Group><Rule id="1"></Rule><Rule id="2"/></Group><Rule id="3"/><RuleResult/></Benchmark>`)
	assert.Equal(t, 3, CountXMLElements(t, xml, "Rule"), "counts open <Rule> tags, not </Rule> or <RuleResult>")
	assert.Equal(t, 0, CountXMLElements(t, xml, "rule-result"))

	rr := []byte(`<a><rule-result/><rule-result/></a>`)
	assert.Equal(t, 2, CountXMLElements(t, rr, "rule-result"), "hyphenated element names")
}

// A non-UTF-8 encoding declaration (e.g. veracode.xml is ISO-8859-1) must not
// error the decoder — the CharsetReader passes bytes through since we only count
// ASCII element names.
func TestCountXMLElements_NonUTF8Encoding(t *testing.T) {
	in := []byte(`<?xml version="1.0" encoding="ISO-8859-1"?><a><Rule/><Rule/></a>`)
	assert.Equal(t, 2, CountXMLElements(t, in, "Rule"))
}

func TestCountJSONItemsUnderKey(t *testing.T) {
	j := []byte(`{"groups":[{"controls":[{"id":"a","controls":[{"id":"a.1"}]},{"id":"b"}]}]}`)
	assert.Equal(t, 3, CountJSONItemsUnderKey(t, j, "controls"), "counts nested controls at any depth: a, b, a.1")
	assert.Equal(t, 0, CountJSONItemsUnderKey(t, j, "missing"))
}

func TestTotalRequirementsBothShapes(t *testing.T) {
	results := map[string]interface{}{
		"baselines": []interface{}{
			map[string]interface{}{"requirements": []interface{}{struct{}{}, struct{}{}}},
		},
	}
	assert.Equal(t, 2, TotalRequirements(t, results), "HDFResults: baselines[].requirements")

	baseline := map[string]interface{}{"requirements": []interface{}{struct{}{}, struct{}{}, struct{}{}}}
	assert.Equal(t, 3, TotalRequirements(t, baseline), "HDFBaseline: top-level requirements")
}

func TestAssertRequirementCount(t *testing.T) {
	baseline := map[string]interface{}{"requirements": []interface{}{struct{}{}, struct{}{}}}
	AssertRequirementCount(t, baseline, 2, "happy path")
}

func TestAssertOverrideCount(t *testing.T) {
	amendments := map[string]interface{}{"overrides": []interface{}{struct{}{}, struct{}{}, struct{}{}}}
	assert.Equal(t, 3, TotalOverrides(t, amendments))
	AssertOverrideCount(t, amendments, 3, "override count")
}

File: hdf-converters/shared/typescript/anchor.test.ts
import { describe, it, expect } from 'vitest';
import {
  countXmlElements,
  countJsonItemsUnderKey,
  assertRequirementCount,
  assertOverrideCount,
} from './anchor.js';

describe('anchor: countXmlElements', () => {
  it('counts opening tags, ignoring closing, namespaced, and substring tags', () => {
    const xml =
      '<Benchmark><Group><Rule id="1"/><Rule id="2"></Rule></Group>' +
      '<xccdf:Rule id="3"/><RuleResult/></Benchmark>';
    // 2 plain + 1 namespaced <Rule>; <RuleResult> and </Rule> excluded.
    expect(countXmlElements(xml, 'Rule')).toBe(3);
  });

  it('counts hyphenated element names', () => {
    expect(countXmlElements('<a><rule-result/><rule-result/></a>', 'rule-result')).toBe(2);
  });

  it('returns 0 when the element is absent', () => {
    expect(countXmlElements('<a><b/></a>', 'Rule')).toBe(0);
  });
});

describe('anchor: countJsonItemsUnderKey', () => {
  it('counts array items under a key at any depth, including nested', () => {
    const json = JSON.stringify({
      groups: [{ controls: [{ id: 'a', controls: [{ id: 'a.1' }] }, { id: 'b' }] }],
    });
    expect(countJsonItemsUnderKey(json, 'controls')).toBe(3); // a, b, a.1
  });

  it('returns 0 when the key is absent', () => {
    expect(countJsonItemsUnderKey('{"x":1}', 'controls')).toBe(0);
  });
});

describe('anchor: assertRequirementCount', () => {
  it('passes on the HDFResults shape (baselines[].requirements)', () => {
    assertRequirementCount(
      JSON.stringify({ baselines: [{ requirements: [{}, {}] }] }),
      2,
      'results shape',
    );
  });

  it('passes on the HDFBaseline shape (top-level requirements)', () => {
    assertRequirementCount({ requirements: [{}, {}, {}] }, 3, 'baseline shape');
  });

  it('throws when want is 0 (an anchor with want=0 proves nothing)', () => {
    expect(() => assertRequirementCount({ requirements: [] }, 0, 'x')).toThrow();
  });

  it('throws on a count mismatch', () => {
    expect(() => assertRequirementCount({ requirements: [{}] }, 2, 'x')).toThrow();
  });
});

describe('anchor: assertOverrideCount', () => {
  it('passes on matching overrides[] count (string or object)', () => {
    assertOverrideCount(JSON.stringify({ overrides: [{}, {}, {}] }), 3, 'string input');
    assertOverrideCount({ overrides: [{}, {}] }, 2, 'object input');
  });

  it('throws on a count mismatch', () => {
    expect(() => assertOverrideCount({ overrides: [{}] }, 2, 'x')).toThrow();
  });
});
